plugins/inputs/xtremio/xtremio.go

```diff
@@ -0,0 +1,485 @@
+				// refactor opportunity would be for the intial collector goroutines to
+				// return the results while exiting the goroutine, and then a series of
+				// goroutines can be kicked off for the subqueries. That way there is no
+				// nesting of goroutines.
+				switch collector {
+				case "bbus":
+					wg.Add(1)
+					go xio.gatherBBUs(acc, url, &wg)
+				case "clusters":
+					wg.Add(1)
+					go xio.gatherClusters(acc, url, &wg)
+				case "ssds":
+					wg.Add(1)
+					go xio.gatherSSDs(acc, url, &wg)
+				case "volumes":
+					wg.Add(1)
+					go xio.gatherVolumes(acc, url, &wg)
+				case "xms":
+					wg.Add(1)
+					go xio.gatherXMS(acc, url, &wg)
+				default:
+					acc.AddError(fmt.Errorf("specified collector %q isn't supported", collector))
+				}
+			}
+		}(collector)
+	}
+	wg.Wait()
+
+	xio.cookie = nil
+
+	return nil
+}
+
+func (xio *XtremIO) gatherBBUs(acc telegraf.Accumulator, url string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := xio.call(url)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	data := BBU{}
+	err = json.Unmarshal([]byte(resp), &data)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	tags := map[string]string{
+		"serial_number": data.Content.Serial,
+		"guid":          data.Content.Guid,
+		"power_feed":    data.Content.PowerFeed,
+		"name":          data.Content.Name,
+		"model_name":    data.Content.ModelName,
+	}
+	fields := make(map[string]interface{})
+	fields["bbus_power"] = data.Content.BBUPower
+	fields["bbus_average_daily_temp"] = data.Content.BBUDailyTemp
+	fields["bbus_enabled"] = (data.Content.BBUEnabled == "enabled")
+	fields["bbus_ups_need_battery_replacement"] = (data.Content.BBUNeedBat == "true")
+	fields["bbus_ups_low_battery_no_input"] = (data.Content.BBULowBat == "true")
+
+	acc.AddFields("xio", fields, tags)
+}
+
+func (xio *XtremIO) gatherClusters(acc telegraf.Accumulator, url string, wg *sync.WaitGroup) {
+	defer wg.Done()
+	resp, err := xio.call(url)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	data := Clusters{}
+	err = json.Unmarshal([]byte(resp), &data)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+
+	tags := map[string]string{
+		"hardware_platform":      data.Content.HardwarePlatform,
+		"license_id":             data.Content.LicenseId,
+		"guid":                   data.Content.Guid,
+		"name":                   data.Content.Name,
+		"sys_psnt_serial_number": data.Content.SerialNumber,
+	}
+	fields := make(map[string]interface{})
+	fields["clusters_compression_factor"] = data.Content.CompressionFactor
+	fields["clusters_percent_memory_in_use"] = data.Content.MemoryUsed
+	fields["clusters_read_iops"] = data.Content.ReadIops
+	fields["clusters_write_iops"] = data.Content.WriteIops
+	fields["clusters_number_of_volumes"] = data.Content.NumVolumes
+	fields["clusters_free_ssd_space_in_percent"] = data.Content.FreeSSDSpace
+	fields["clusters_ssd_num"] = data.Content.NumSSDs
+	fields["clusters_data_reduction_ratio"] = data.Content.DataReductionRatio
```

Comment:
The fields can be initialized the same way as the tags (see the example above). This makes it easier to read. Don't forget to call `make fmt` at the end to get a nice alignment.